File: Bulwark.Core/Engine/EgressRateMonitor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Bulwark.Core.Models;

namespace Bulwark.Core.Engine;

/// <summary>
/// 外联速率/扇出监视器(独创·有状态时序检测)。
///
/// 补全网络防护的第三条腿(前两条:<see cref="BeaconDetector"/> 时序节律、
/// <see cref="DgaDomainAnalyzer"/> 域名随机度)。本监视器关注**外联的"量"与"面"**:
///   · 速率突发(rate burst):单进程在极短时间内发起远超常态的外联次数;
///   · 目标扇出(fan-out):单进程在短时间内连向大量【不同】远端目标。
///
/// 这两类是端口扫描 / 内网横移 / 蠕虫传播 / C2 多节点轮询 / 数据外传分块上传的共同特征。
/// 正常程序的外联通常集中在少数几个固定服务端(浏览器除外),很少在数秒内打向几十个不同 IP。
///
/// 关键约束(遵循全局低误报原则):速率/扇出是**软信号**,
/// <b>单独不得</b>置位 <see cref="SecurityEvent.HasThreatIndicator"/> 或直接处置 ——
/// 浏览器、下载器、P2P、爬虫、更新检查都可能高速多目标外联。
/// 它只贡献 RiskScore / RiskReasons,由调用方在与另一硬指标(信标 / DGA 升格 /
/// 未签名可疑目录主体)共现时才升格(互证机制)。
///
/// 线程安全:单锁串行化,带容量/过期治理。
/// </summary>
public sealed class EgressRateMonitor
{
    private readonly object _gate = new();
    private readonly Dictionary<int, ProcState> _byPid = new();
    private readonly TimeSpan _window;
    private readonly int _maxPids;

    /// <summary>窗口内"外联次数"达到此值视为速率突发。</summary>
    private readonly int _rateThreshold;

    /// <summary>窗口内"不同远端目标数"达到此值视为高扇出。</summary>
    private readonly int _fanoutThreshold;

    public EgressRateMonitor(
        TimeSpan? window = null,
        int rateThreshold = 40,
        int fanoutThreshold = 20,
        int maxPids = 4096)
    {
        _window = window ?? TimeSpan.FromSeconds(10);
        _rateThreshold = Math.Max(10, rateThreshold);
        _fanoutThreshold = Math.Max(8, fanoutThreshold);
        _maxPids = Math.Max(64, maxPids);
    }

    /// <summary>
    /// 观测一次网络外联事件,更新进程状态并研判速率/扇出异常。
    /// 仅对 <see cref="EventType.NetworkConnect"/> 有意义,其它类型返回 0 分。
    /// </summary>
    public (int Score, List<string> Reasons) Observe(SecurityEvent e)
    {
        var reasons = new List<string>();
        if (e is null || e.Type != EventType.NetworkConnect || e.ActorPid <= 0)
            return (0, reasons);

        string remote = NormalizeRemote(e.Target);
        var now = e.TimestampUtc == default ? DateTime.UtcNow : e.TimestampUtc;
        int score = 0;

        lock (_gate)
        {
            if (!_byPid.TryGetValue(e.ActorPid, out var st))
            {
                st = new ProcState();
                _byPid[e.ActorPid] = st;
            }
            st.Add(remote, now, _window);

            int count = st.CountInWindow(now, _window);
            int distinct = st.DistinctTargetsInWindow(now, _window);

            // 1) 速率突发
            if (count >= _rateThreshold)
            {
                int over = count - _rateThreshold;
                score += 25 + Math.Min(over, 40);
                reasons.Add($"{_window.TotalSeconds:0}秒内高速外联 {count} 次(疑似扫描/外传/C2 轮询)");
            }

            // 2) 目标扇出(更强信号:连大量不同目标)
            if (distinct >= _fanoutThreshold)
            {
                int over = distinct - _fanoutThreshold;
                score += 30 + Math.Min(over * 2, 40);
                reasons.Add($"{_window.TotalSeconds:0}秒内连向 {distinct} 个不同目标(疑似横移/扫描/蠕虫传播)");
            }

            EvictIfNeeded(now);
        }

        return (Math.Min(score, 100), reasons);
    }

    /// <summary>进程退出时清理其状态(可选)。</summary>
    public void Forget(int pid)
    {
        lock (_gate) _byPid.Remove(pid);
    }

    /// <summary>当前跟踪进程数(诊断/测试用)。</summary>
    public int TrackedProcessCount
    {
        get { lock (_gate) return _byPid.Count; }
    }

    private void EvictIfNeeded(DateTime now)
    {
        var stale = now - TimeSpan.FromTicks(_window.Ticks * 4);
        List<int>? dead = null;
        foreach (var kv in _byPid)
            if (kv.Value.LastUtc < stale)
                (dead ??= new List<int>()).Add(kv.Key);
        if (dead is not null)
            foreach (var pid in dead) _byPid.Remove(pid);

        if (_byPid.Count > _maxPids)
        {
            var oldest = _byPid.OrderBy(kv => kv.Value.LastUtc)
                .Take(_byPid.Count - _maxPids)
                .Select(kv => kv.Key).ToList();
            foreach (var pid in oldest) _byPid.Remove(pid);
        }
    }

    private static string NormalizeRemote(string? target)
    {
        if (string.IsNullOrEmpty(target)) return "?";
        var t = target.Trim().ToLowerInvariant();
        int colon = t.LastIndexOf(':');
        if (colon > 0 && colon < t.Length - 1 && t.Skip(colon + 1).All(char.IsDigit))
            t = t.Substring(0, colon);
        return t;
    }

    /// <summary>单进程的外联滑动窗口状态。</summary>
    private sealed class ProcState
    {
        public DateTime LastUtc;
        private readonly List<(string Remote, DateTime At)> _events = new();
        private const int HardCap = 4096;

        public void Add(string remote, DateTime at, TimeSpan window)
        {
            _events.Add((remote, at));
            LastUtc = at;
            var cutoff = at - window;
            int i = 0;
            while (i < _events.Count && _events[i].At < cutoff) i++;
            if (i > 0) _events.RemoveRange(0, i);
            if (_events.Count > HardCap)
                _events.RemoveRange(0, _events.Count - HardCap);
        }

        public int CountInWindow(DateTime now, TimeSpan window)
        {
            var cutoff = now - window;
            int n = 0;
            foreach (var ev in _events) if (ev.At >= cutoff) n++;
            return n;
        }

        public int DistinctTargetsInWindow(DateTime now, TimeSpan window)
        {
            var cutoff = now - window;
            return _events.Where(ev => ev.At >= cutoff)
                          .Select(ev => ev.Remote)
                          .Distinct()
                          .Count();
        }
    }
}

File: cpp/shared/src/engine/EgressRateMonitor.cpp
#include "bulwark/engine/EgressRateMonitor.h"
#include <QSet>
#include <algorithm>

namespace bulwark::engine {
using detail::u;

namespace {
constexpr int kHardCap = 4096;

QString normalizeRemote(const QString& target) {
    if (target.isEmpty()) return QStringLiteral("?");
    QString t = target.trimmed().toLower();
    const int colon = t.lastIndexOf(QLatin1Char(':'));
    if (colon > 0 && colon < t.size() - 1) {
        bool allDigits = true;
        for (int i = colon + 1; i < t.size(); ++i)
            if (!t.at(i).isDigit()) { allDigits = false; break; }
        if (allDigits) t = t.left(colon);
    }
    return t;
}
} // namespace

void EgressRateMonitor::ProcState::add(const QString& remote, const QDateTime& at, int windowSecs) {
    events.append({ remote, at });
    lastUtc = at;
    const QDateTime cutoff = at.addSecs(-windowSecs);
    int i = 0;
    while (i < events.size() && events.at(i).second < cutoff) ++i;
    if (i > 0) events.remove(0, i);
    if (events.size() > kHardCap) events.remove(0, events.size() - kHardCap);
}

int EgressRateMonitor::ProcState::countInWindow(const QDateTime& now, int windowSecs) const {
    const QDateTime cutoff = now.addSecs(-windowSecs);
    int n = 0;
    for (const auto& ev : events) if (ev.second >= cutoff) ++n;
    return n;
}

int EgressRateMonitor::ProcState::distinctTargetsInWindow(const QDateTime& now, int windowSecs) const {
    const QDateTime cutoff = now.addSecs(-windowSecs);
    QSet<QString> set;
    for (const auto& ev : events) if (ev.second >= cutoff) set.insert(ev.first);
    return set.size();
}

EgressRateMonitor::EgressRateMonitor(int windowSeconds, int rateThreshold, int fanoutThreshold, int maxPids) {
    windowSecs_ = windowSeconds > 0 ? windowSeconds : 10;
    rateThreshold_ = qMax(10, rateThreshold);
    fanoutThreshold_ = qMax(8, fanoutThreshold);
    maxPids_ = qMax(64, maxPids);
}

ScoreResult EgressRateMonitor::observe(const SecurityEvent& e) {
    ScoreResult r;
    if (e.type != EventType::NetworkConnect || e.actorPid <= 0) return r;

    const QString remote = normalizeRemote(e.target);
    const QDateTime now = e.timestampUtc.isValid() ? e.timestampUtc : QDateTime::currentDateTimeUtc();
    int score = 0;

    QMutexLocker locker(&gate_);
    ProcState& st = byPid_[e.actorPid];
    st.add(remote, now, windowSecs_);

    const int count = st.countInWindow(now, windowSecs_);
    const int distinct = st.distinctTargetsInWindow(now, windowSecs_);

    if (count >= rateThreshold_) {
        const int over = count - rateThreshold_;
        score += 25 + qMin(over, 40);
        r.reasons << (QString::number(windowSecs_) + u("秒内高速外联 ") + QString::number(count) +
                      u(" 次(疑似扫描/外传/C2 轮询)"));
    }
    if (distinct >= fanoutThreshold_) {
        const int over = distinct - fanoutThreshold_;
        score += 30 + qMin(over * 2, 40);
        r.reasons << (QString::number(windowSecs_) + u("秒内连向 ") + QString::number(distinct) +
                      u(" 个不同目标(疑似横移/扫描/蠕虫传播)"));
    }

    evictIfNeeded(now);
    r.score = qMin(score, 100);
    return r;
}

void EgressRateMonitor::forget(int pid) {
    QMutexLocker locker(&gate_);
    byPid_.remove(pid);
}

int EgressRateMonitor::trackedProcessCount() {
    QMutexLocker locker(&gate_);
    return byPid_.size();
}

void EgressRateMonitor::evictIfNeeded(const QDateTime& now) {
    const QDateTime stale = now.addSecs(-static_cast<qint64>(windowSecs_) * 4);
    for (auto it = byPid_.begin(); it != byPid_.end(); ) {
        if (it.value().lastUtc < stale) it = byPid_.erase(it);
        else ++it;
    }
    if (byPid_.size() > maxPids_) {
        QVector<QPair<QDateTime, int>> byAge;
        byAge.reserve(byPid_.size());
        for (auto it = byPid_.constBegin(); it != byPid_.constEnd(); ++it)
            byAge.append({ it.value().lastUtc, it.key() });
        std::sort(byAge.begin(), byAge.end(),
                  [](const auto& a, const auto& b) { return a.first < b.first; });
        const int toRemove = byPid_.size() - maxPids_;
        for (int i = 0; i < toRemove; ++i) byPid_.remove(byAge.at(i).second);
    }
}

} // namespace bulwark::engine
